pr_docker_library.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+
+import datetime
+import difflib
+import git  # gitpython
+import github
+import os
+import shutil
+import tempfile
+
+
+# Path to root of this repo
+BASE_PATH = os.path.dirname(os.path.abspath(__file__))
+# Path to official docker library repo
+DOCKER_LIBRARY_REPO_URL = 'git@github.com:docker-library/official-images.git'
+DOCKER_LIBRARY_REPO = 'docker-library/official-images'
+# Path to fork to update with docker images
+FORK_LIBRARY_REPO_URL = 'git@github.com:ros-infrastructure/official-images.git'
+FORK_LIBRARY_REPO_USER = 'ros-infrastructure'
+
+# Token to use to interact with github
+GIT_USER = os.environ.get('GITHUB_USER','')
+GIT_EMAIL = os.environ.get('GITHUB_EMAIL','')
+GIT_TOKEN = os.environ.get('GITHUB_TOKEN','')
+GIT_AUTHOR = "{user} <{email}>".format(user=GIT_USER, email=GIT_EMAIL)
+GIT_TOKEN = os.environ.get('GITHUB_TOKEN','')
+
+
+PR_TEMPLATE = """
+# Update library definition for '{name}'
+
+<details><summary>Diff</summary>
+
+```diff
+{diff}
+```
+</details>
+"""
+
+
+def read_local_library_definition(name):
+    # Library definition file for `ros` is in `ros/ros`
+    with open(os.path.join(BASE_PATH, name, name), 'r') as fin:
+        return fin.read()
+
+
+def read_upstream_library_definition(name):
+    # Get the manifest from the upstream repo 
+    with tempfile.TemporaryDirectory(prefix='osrf-docker-images') as temp_dir:
+        git.Repo.clone_from(DOCKER_LIBRARY_REPO_URL, temp_dir, depth=1) 
+        with open(os.path.join(temp_dir, 'library', name)) as fin:
+            return fin.read()
+
+
+def pr_with_sentinel_exists(sentinel):
+    gh_handle = github.Github(GIT_TOKEN)
+    repo = gh_handle.get_repo(DOCKER_LIBRARY_REPO)
+    for pr in repo.get_pulls(state='open'):
+        if sentinel in pr.body:
+            return True
+    return False
+
+
+def diff_library_definitions(local_ldf, upstream_ldf):
+    differ = difflib.Differ()
+    diff = differ.compare(
+        upstream_ldf.splitlines(keepends=True),
+        local_ldf.splitlines(keepends=True))
+    diff = ''.join(diff)
+    return diff
+
+
+def update_library_definition_fork(name):
+    # Create branch on fork that is up to date with upstream
```

Comment:
Is this pushing the manifest (post merged) copied from osrf/docker_images to the `FORK_LIBRARY_REPO_URL` or only that from the checked out PR? To consolidate all the release changes for a given docker repo to one at a time, I think pushing the merged state of the repo manifest would be less overwhelming upstream that PRing each change. E.g. we'd only approve this automated upstream PR for the last merged PR to a given docker repo to batch release changes. This is similar to what we do now manually.

Reply:
I don't understand the question. Here is an example PR created by this script: ros-infrastructure/official-images#8 . Does the diff answer it?

Reply:
Your pr is a multiple merged distro changes, yet local PRs by the @osrf-docker-builder separate each distro PR, so the maintainer should be prudent in waiting for all the changes targeting a given repo manifest to be merged before upstreaming the final updated manifest. To avoid opening to many PR, they are batched into one periodic PR for that respective dockerhub repo.

Example PR:
https://github.com/osrf/docker_images/pull/233
https://github.com/osrf/docker_images/pull/234
https://github.com/osrf/docker_images/pull/235
https://github.com/osrf/docker_images/pull/236
https://github.com/osrf/docker_images/pull/231

Upstream PR:
https://github.com/docker-library/official-images/pull/5573